accepted/single-file/design.md

```diff
@@ -66,225 +32,254 @@ Publishing to a single file can be triggered by adding the following property to
+* For Linux mini-dumps: `createdump`, `libmscordaccore.so` 
+* For ETW / LTTng: `clretwrc.dll`, `libcoreclrtraceptprovider.so` 
+* For exception stack trace source information, error string resources on Windows: `Microsoft.DiaSymReader.Native.amd64.dll`, `mscorrc.debug.dll`, `mscorrc.dll`
+
+When targeting `win7` platform, several additional DLLs are necessary (`api-*.dll`) to handle API compatibility. These files must be alongside the AppHost for the app to even start execution.
+Therefore, we propose that targeting `win7-*` should not be supported when publishing apps as a single-file.
+
+## The Runtime
+
+### Dependency Resolution
+
+In order to resolve assemblies, the embedded resources are probed first, followed by other probing paths encoded in `TRUSTED_PLATFORM_ASSEMBLIES`. 
+
+### PEImage loader
+
+* IL assemblies are loaded directly from the bundle.
+    * The portion of the single-file bundle containing the required assembly is memory mapped, and the contents are appropriately interpreted by the runtime.
+
+* ReadyToRun Assemblies
+    * On Linux, ReadyToRun assemblies are loaded directly from bundle. The various sections in the PE file are mapped at appropriate addresses, and offsets are fixed up.
+    * On MAC, ReadyToRun assemblies are loaded similar to Linux. However, Mojave hardened runtime doesn't allow executable mappings of a file. Therefore, the contents of an assembly are read from the bundle into pre-allocated executable memory.
+    * On Windows, due to certain limitations in memory mapping routines described below, ReadyToRun assemblies are loaded by memory mapping the file and copying sections to appropriate offsets.
+
+* ReadyToRun [Composite](https://github.com/dotnet/runtime/blob/master/docs/design/features/readytorun-composite-format-design.md) Assemblies are expected to be loaded similar to ReadyToRun assemblies. 
+
+#### Windows Limitations
+The Windows mapping routines have the following limitations:
+* [CreateFileMapping](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createfilemappinga) has no option to create a mapping for a part of the file (no offset argument).
+    This means that:
+    * We cannot use the (SEC_IMAGE) attribute to perform automatic section-wise loading (circumventing alignment requirements) of bundled assemblies directly.
+    * Instead we need to map each section independently.
+* [MapViewOfFile]( https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-mapviewoffileex) can only map parts of a file aligned at [memory allocation granularity]( https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/ns-sysinfoapi-system_info), which is 64KB.
+    This means that:
+    * Each section within the assemblies should be aligned at 64KB – which is not guaranteed by the crossgen compiler.
+    * In order to memory-map one embedded assembly at a time, the assemblies in the bundle must be aligned at 64KB boundaries.
```

Comment:
The runtime can be fixed to handle assemblies that do not start at 64kB boundaries.

Reply:
I am assuming that if we don't change the runtime, the binary size would be bloated by padding up to the 64k boundary for each assembly in the app/framework? 
That would run counter to the idea to be able to create smaller single-file apps.

Reply:
> The runtime can be fixed to handle assemblies that do not start at 64kB boundaries.

Yes, I agree that the runtime can round-down to a 64KB boundary, and read the headers from an appropriate offset. But after the headers are processed, the individual sections themselves need to be 64 KB aligned. So, it is probably sufficient to state the limitation about section alignment.

Reply:
> I am assuming that if we don't change the runtime, the binary size would be bloated by padding up to the 64k boundary for each assembly in the app/framework?
> That would run counter to the idea to be able to create smaller single-file apps.

The current proposal doesn't involve 64KB alignment, but an extra copy step at run-time. 

The 64 KB section alignment is a Windows limitation. So the suggested mitigation (likely as a subsequent step) is to use a composite R2R with aligned sections.

Reply:
> But after the headers are processed, the individual sections themselves need to be 64 KB aligned.

They do not need to be. We have full control over the embedded binary format. We do not have to embed the exact copy of the PE file:
- We can reformat the sections as we embed them, so that we can map the whole file into memory at once, and just flip the permissions on pages as necessary.
- Also, for IL-only (ie no R2R), the runtime can use the flat mapping (we do that on Unix).

Reply:
In other words, this is not a fundamental limitation. This is limited by the amount of effort we can afford to put into this.